build/patches/Add-Viewport-Protection.patch

```diff
@@ -0,0 +1,667 @@
++  return screen_info.available_rect.width() * GetScaleOverride(frame->GetPage());
+ }
+ 
+ void Screen::Trace(Visitor* visitor) const {
+diff --git a/third_party/blink/renderer/core/frame/visual_viewport.cc b/third_party/blink/renderer/core/frame/visual_viewport.cc
+--- a/third_party/blink/renderer/core/frame/visual_viewport.cc
++++ b/third_party/blink/renderer/core/frame/visual_viewport.cc
+@@ -467,6 +467,10 @@ void VisualViewport::SetScale(float scale) {
+                       gfx::PointAtOffsetFromOrigin(offset_));
+ }
+ 
++void VisualViewport::SetPageWidthOverride(int value) {
++  page_width_override_ = value;
++}
++
+ double VisualViewport::OffsetLeft() const {
+   // Offset{Left|Top} and Width|Height are used by the DOMVisualViewport to
+   // expose values to JS. We'll only ever ask the visual viewport for these
+diff --git a/third_party/blink/renderer/core/frame/visual_viewport.h b/third_party/blink/renderer/core/frame/visual_viewport.h
+--- a/third_party/blink/renderer/core/frame/visual_viewport.h
++++ b/third_party/blink/renderer/core/frame/visual_viewport.h
+@@ -145,6 +145,9 @@ class CORE_EXPORT VisualViewport : public GarbageCollected<VisualViewport>,
+   float Scale() const { return scale_; }
+   bool IsPinchGestureActive() const { return is_pinch_gesture_active_; }
+ 
++  void SetPageWidthOverride(int);
++  int PageWidthOverride() const { return page_width_override_; }
++
+   // Convert the given rect in the main LocalFrameView's coordinates into a rect
+   // in the viewport. The given and returned rects are in CSS pixels, meaning
+   // scale isn't applied.
+@@ -382,6 +385,8 @@ class CORE_EXPORT VisualViewport : public GarbageCollected<VisualViewport>,
+   CompositorElementId elasticity_effect_node_id_;
+ 
+   bool needs_paint_property_update_;
++
++  int page_width_override_;
+ };
+ 
+ }  // namespace blink
+diff --git a/third_party/blink/renderer/core/html/html_meta_element.cc b/third_party/blink/renderer/core/html/html_meta_element.cc
+--- a/third_party/blink/renderer/core/html/html_meta_element.cc
++++ b/third_party/blink/renderer/core/html/html_meta_element.cc
+@@ -22,6 +22,7 @@
+ 
+ #include "third_party/blink/renderer/core/html/html_meta_element.h"
+ 
++#include "base/rand_util.h"
+ #include "third_party/blink/public/mojom/frame/color_scheme.mojom-blink.h"
+ #include "third_party/blink/renderer/core/css/style_engine.h"
+ #include "third_party/blink/renderer/core/dom/document.h"
+@@ -565,6 +566,27 @@ void HTMLMetaElement::ProcessContent() {
+   if (!IsInDocumentTree())
+     return;
+ 
++  bool process_default = true;
++
++  if (GetDocument().GetFrame()) {
++    blink::LocalFrame* frame = GetDocument().GetFrame();
++    blink::WebContentSettingsClient* settings = frame->GetContentSettingsClient();
++    blink::Page* page = frame->GetPage();
++
++    if (settings->AllowViewportChange(false)) {
++      if (page->PageWidthOverride() == 0) {
++        page->SetPageWidthOverride(base::RandInt(-5, 15));
```

Comment:
this is where the random scale is decided.
the factor I have set for now is -5% to +15% compared to the current size.
we probably need to check if it is a correct range.

Reply:
`-5` was too annoying, I changed to `-3 ... +15` with a factor of 10